Feature assemblies base

webdevinition opened the pull request:
Hello @jbtronics,

In relation to https://github.com/Part-DB/Part-DB-server/pull/1051 the feature for adding the base of assembly as an new data source.
Also mentioned in the discussion under https://github.com/Part-DB/Part-DB-server/discussions/1041, section 1.

The changes are the remaining delta to the other extracted feature pull requests related to the mentioned discussion.

<img width="3172" height="1503" alt="2025-10-17 Assembly BOM configuration" src="https://github.com/user-attachments/assets/b63a98ad-2a45-4f4d-a34f-17ce044772ba" />

_Configuration opportunity of Bills of Materials (BOMs) with referencing of other assemblies, cycle prevention included_

<img width="1785" height="4200" alt="2025-10-17-Import-assembly oppurtunities" src="https://github.com/user-attachments/assets/41a39ac7-c39a-4f9e-b2e9-885ef85ff5d1" />

_Simplified import overview with template description for filling the parts lists (BOM) of an assembly_

<img width="3173" height="768" alt="2025-10-17-Human readable export" src="https://github.com/user-attachments/assets/dde50794-b964-47b2-962e-5b2f475078f7" />

_Human-readable export in CSV and PDF format for simplified viewing of basic information_

<img width="1011" height="960" alt="2025-10-17-Human readable export - PDF example" src="https://github.com/user-attachments/assets/05a8ce4c-d8a3-4a80-8a7a-a1ed4c25e410" />

_Human-readable export in PDF format to simplify understanding of which parts are actually required for a multiply nested assembly, for example, completely independent of a build function as is the case with projects._

Thanks for everything so far! I am grateful for integration!

Best regards,
Marcel

Commit:
Füge Baugruppen hinzu

Commit:
Leerzeilen beim Import überspringen

Commit:
AssemblyBomEntriesDataTable: Sortierfelder für Beschreibung und Designator ergänzen

Commit:
Assembly-Filter und -Tab ergänzen

Fügt Unterstützung für Assembly-Filter in der Suche sowie ein neues Tab für Assemblies in der Filteransicht hinzu. Anpassungen beinhalten Änderungen an Datenbankabfragen, Formularen und Templates.

Commit:
Assembly-BOM-Einträge und Übersetzungen aktualisieren

Verbessert die Präzision der Übersetzungen für 'assembly_bom_entry' in mehreren Sprachen. Ergänzt neue Label- und Pluralformen in den Dateien sowie zugehörige Zuordnungen in Klassen.

Commit:
EntityExporter: Konstruktor-Definition bereinigen

Commit:
EntityExporter: Konstruktor-Deklaration korrigieren; Dienstkonfiguration für Tests aktualisieren

Fügt Standardwerte für autowire und autoconfigure in der Testumgebung hinzu. Markiert EntityExporter als öffentlich und bereinigt die Konstruktor-Definition in der zugehörigen Klasse.

Commit:
services.yaml: AssemblyPartAggregator registrieren

jbtronics commented:
Im still not really convinced that introducing a new kind of entity is the best approach there. Especially as this PR is very big and complex (136 file changes) and a lot of logic seems to be duplicated from projects.

Maybe the goal could also be reached by adding some new functionality to the existing projects (like adding a flag, that makes them non-buildable and only work as generic placeholder, or assembly if you wanna call it that way).

I am not sure how many people would actually profit from that feature anyway, and if it is worth to the additional complexity/maintenance effort of the PR in the current form.

Commit:
Unterstützung für Projekt- und Baugruppensuche zum QuickSearch-Suggest hinzufügen

Commit:
Statistik-Bereich um Tab für Projekte/Baugruppen erweitern

Commit:
Cleanup-Logik für Baugruppen und BOM-Einträge im Statistik-Bereich überarbeiten bzw. erweitern

Commit:
statistics_assembly_controller bzgl. Alert-Ausgabe aktualisieren

Commit:
Typdefinitionen in BOMImporter anpassen

Commit:
Übergabe an stringToBOMEntries in BOMImporter anpassen

Commit:
Überprüfung auf Tabellenexistenz in Migration hinzufügen.

Commit:
Datasource-spezifische Suche für Projects/Assemblies sowie Parts umsetzen

Commit:
Unterstützung für IPN-Suche in der BOM-Konfiguration bei Part- und Assembly-Auswahl hinzufügen.
JavaScript BOM-Synchronisation (nur ein Part oder ein Assembly wählbar) ergänzen.
laceholder für Name-Angabe setzen: Entweder Part- oder Assembly-Name.

Commit:
Status- und Herstellerinformationen in Readable Assembly-Exports hinzufügen.

Commit:
Turbo-Kompatibilität verbessern.

Hinzufügen von Prüfungen auf Vorhandensein von DOM-Elementen in mehreren Controllern sowie Optimierung der Form-Submit-Logik für bessere Turbo-Integration. Anpassung von Template-Elementen und JavaScript-Events zur Unterstützung von Turbo Morphing auf der Startseite.

Commit:
Fehlerbehandlung für Update-Checker verbessern.

lukas-runge commented:
I'd also love to have a feature like this. Doesn't need to be a new thing entirely - I can totally see this work on the current projects structure, but it would be great if I'd be able to define an "assembly of projects" in defined quantities so I can see (when pressing "Build") if I have all the parts in stock to build all of the projects at their desired quantities or not. The usecase is basically any project with more than one PCB inside. My current Project has 11 different designs contained within. The only current workaround for something like this is to merge all boms, respecting quantities of every bom and creating an additional new project with all of it inside. Ofc I can understand the hesitation towards this obviously vibe-coded prototype. But still think the goal behind this is worth pursuing. Imo the MVP of this would just allow to add projects as parts to other projects bom. This way a product could e.g. have 20 of the same board projects inside and 5 of another kind and then when pressing "Build" on the product project which nests all other component projects inside this would then recursively get all parts and calculate a sum to tell me what I am missing to build the entire thing. What do you think @jbtronics? Would this be viable to build without a headache? 🙌

Commit:
Aggregationslogik in AssemblyPartAggregator entfernen. Sub-Assembly Rows ebenfalls via processBomEntriesWithAggregatedParts ermitteln
